modyn/config/schema/pipeline/evaluation/metrics.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import annotations
+
+from typing import Annotated, Any, Callable, Literal, Union
+
+from modyn.config.schema.base_model import ModynBaseModel
+from modyn.utils.utils import EVALUATION_TRANSFORMER_FUNC_NAME, deserialize_function
+from pydantic import BaseModel, Field, field_validator, model_validator
+
+
+class _BaseMetricConfig(ModynBaseModel):
+    config: dict[str, Any] = Field({}, description="Configuration for the evaluation metric.")
+    evaluation_transformer_function: str | None = Field(
+        None,
+        description="A function used to transform the model output before evaluation.",
+    )
+
+    @field_validator("evaluation_transformer_function", mode="before")
+    @classmethod
+    def validate_evaluation_transformer_function(cls, value: str) -> str | None:
+        if not value:
+            return None
+        try:
+            deserialize_function(value, EVALUATION_TRANSFORMER_FUNC_NAME)
+        except AttributeError as exc:
+            raise ValueError("Function 'evaluation_transformer_function' could not be parsed!") from exc
+        return value
+
+    @property
+    def evaluation_transformer_function_deserialized(self) -> Callable | None:
+        if self.evaluation_transformer_function:
+            return deserialize_function(self.evaluation_transformer_function, "evaluation_transformer_function")
+        return None
+
+    @property
+    def shape_check(self) -> bool:
+        return True
+
+
+class AccuracyMetricConfig(_BaseMetricConfig):
+    name: Literal["Accuracy"] = Field("Accuracy")
+    topn: int = Field(1, description="The top-n accuracy to be computed.", ge=1)
+
+    @property
+    def shape_check(self) -> bool:
+        return self.topn is None
```

Comment:
Shouldn't it be > 1? I did not check this, but since it defaults to 1, can it ever be None? Or is this done before defaults are applied?

Reply:
I am not fully sure I get your comment yet. The logic you posted is indeed the old logic. However, in the pydantic implementation, self.topn is never None, is it? Because self is referring to the accuracy config object, where topn defaults to 1. Hence, right now shape_check is always False, but we want it to be true if topn == 1.

Reply:
Ah, i see, then I misinterpreted the old logic. If one had set `topn=1` manually then `self. _enable_shape_check` would have been False. Let's use the semantics that you just described!